sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/ZE_experimental_bindless_images.asciidoc

```diff
@@ -0,0 +1,368 @@
+Lastly, is `ze_image_mem_alloc_exp_desc_t`.
+This is used for Level Zero to allocate memory and calculate the best pitch size for bindless images.
+
+=== New Structure Types
+
+* enumerator `ZE_BINDLESS_IMAGE_EXP_DESC`
++
+```
+ze_bindless_image_exp_desc_t
+```
+* enumerator `ZE_STRUCTURE_TYPE_IMAGE_MEM_EXP_DESC`
++
+```
+ze_image_mem_exp_desc_t
+```
+* enumerator `ZE_STRUCTURE_TYPE_IMAGE_MEM_ALLOC_DESC`
++
+```
+ze_image_mem_alloc_exp_desc_t
+```
+
+==== ze_bindless_image_exp_desc_t
+
+* `struct ze_bindless_image_exp_desc_t`
++
+Bindless descriptor
+
+Public Members
+
+* `ze_structure_type_t stype`
++
+[in] type of this structure
+* `const void *pNext`
++
+[in][optional] must be null or a pointer to an extension-specific structure (i.e. contains sType and pNext).
+
+`ze_bindless_image_exp_desc_t` is intended to be used to declare an image as bindless.
+If the choice is taken for bindless to be the default then this struct will not be needed.
+In this document, it is assumed that bound images will be the default and bindless images will need to be explicitly declared.
+
+To use this struct it should be passed into `zeImageCreate` to inform Level Zero that the image is to be bindless.
+The returned image handle in turn should have the characteristics to enable bindless images.
+
+==== ze_image_mem_exp_desc_t
+
+* `struct ze_image_mem_exp_desc_t`
++
+Manual image memory descriptor
+
+Public Members
+
+* `ze_structure_type_t stype`
++
+[in] type of this structure
+* `const void *pNext`
++
+[in][optional] must be null or a pointer to an extension-specific structure (i.e. contains sType and pNext).
+*
+`void **ptr`
++
+[in] pointer to device allocation
+* `uint64_t rowPitch`
++
+[in][optional] size of image row pitch in bytes
+* `uint64_t slicePitch`
++
+[in][optional] size of image slice pitch in bytes
+
+`ze_image_mem_exp_desc_t` is intended to allow for the separate memory management of images.
+To use this struct it is to be passed to `zeImageCreate`.
+In addition to passing the pointer to the memory that `zeImageCreate` is to use it also informs Level Zero what pitch the image memory is.
+
+==== ze_image_mem_alloc_exp_desc_t
+
+* `struct ze_image_mem_alloc_exp_desc_t`
++
+Declare pitch of image
+
+Public Members
+
+* `ze_structure_type_t stype`
++
+[in] type of this structure
+* `const void *pNext`
++
+[in][optional] must be null or a pointer to an extension-specific structure (i.e. contains sType and pNext).
+* `uint64_t widthInBytes`
++
+[in] width of image row in bytes
+* `uint64_t height`
++ [in] height of image
+* `uint64_t depth`
++
+[in] depth of image
+* `uint64_t elementSizeBytes`
++
+[in] byte size of single element
+* `uint64_t *rowPitch`
++
+[out][optional] pitch of row
+* `uint64_t *slicePitch`
++
+[out][optional] pitch of slice
+
+`ze_image_mem_alloc_exp_desc_t` is intended to be passed to `zeMemAllocDevice` when manually allocating memory for an image to return optimized pitch information.
```

Comment:
if ze_image_mem_alloc_exp_desc_t will be passed to zeMemAllocDevice, then how would this ensure that size passed to zeMemAllocDevice is enough to fit the image?
What if size would be lower then image requirements ? Is error expected to be returned? 

The below pseudo code assumes that image would fit
width * height * sizeof(float)

This will not be true for all sizes, HW has certain pitch requirements.

Reply:
I believe that there is an oversight in this proposal where we still include the passing of the `size` parameter to `zeMemAllocDevice`. This can cause confusion between the struct we pass which contains the dimensions/sizing and return pitch information, while still passing this `size` parameter. We may need to instead propose a new allocation function.

It was intended that the struct `ze_image_mem_alloc_exp_desc_t` passed to `zeMemAllocDevice` contains the fields `rowPitch` and `slicePitch`, and the function would allocate the appropriate amount of memory with the `widthInBytes`, `height` and `depth` values. These are return variables that the device populates with the optimal pitch values.